src/Database/LSMTree/Internal/MergingTree.hs

```diff
@@ -327,10 +353,170 @@ remainingMergeDebtPendingMerge (PendingMerge _ prs mts) = do
+                >>= leftToRight (flip supplyTree) (toList mt)
+            PendingUnionMerge mts ->
+              splitEqually (flip supplyTree) (V.toList mts) credits
+
+    supplyPreExisting c = \case
+        PreExistingRun _r        -> return c  -- no work to do, all leftovers
+        PreExistingMergingRun mr -> MR.supplyCreditsRelative mr threshold c
+
+    -- supply credits left to right until they are used up
+    leftToRight ::
+         (MR.MergeCredits -> a -> m MR.MergeCredits)
+      -> [a] -> MR.MergeCredits -> m MR.MergeCredits
+    leftToRight _ _      0 = return 0
+    leftToRight _ []     c = return c
+    leftToRight f (x:xs) c = f c x >>= leftToRight f xs
+
+    -- approximately equal, being more precise would require more iterations
+    splitEqually ::
+         (MR.MergeCredits -> a -> m MR.MergeCredits)
+      -> [a] -> MR.MergeCredits -> m MR.MergeCredits
+    splitEqually f xs (MR.MergeCredits credits) =
+        -- first give each tree k = ceil(1/n) credits (last ones might get less).
+        -- it's important we fold here to collect leftovers.
+        -- any remainders go left to right.
+        foldM supplyNth (MR.MergeCredits credits) xs >>= leftToRight f xs
+      where
+        !n = length xs
+        !k = MR.MergeCredits ((credits + (n - 1)) `div` n)
+
+        supplyNth 0 _ = return 0
+        supplyNth c t = do
+            let creditsToSpend = min k c
+            leftovers <- f creditsToSpend t
+            return (c - creditsToSpend + leftovers)
+
+    startPendingMerge reg pm = do
+        (mergeType, rs) <- expectCompletedChildren reg pm
+        assert (V.length rs > 0) $ pure ()
+        runNumber <- uniqueToRunNumber <$> incrUniqCounter uc
+        let runPaths = Paths.runPath (Paths.activeDir root) runNumber
+        mr <-
+          withRollback reg
+            (MR.new hfs hbio resolve runParams mergeType runPaths rs)
+            releaseRef
+        -- no need for the runs anymore, 'MR.new' made duplicates
+        traverse_ (\r -> delayedCommit reg (releaseRef r)) rs
+        return (OngoingTreeMerge mr)
+
+    -- Child references are released using 'delayedCommit', so they get released
+    -- if the whole supply operation runs successfully (so the pending merge
+    -- is replaced).
+    --
+    -- Returned references are registered in the ActionRegistry, so they will
+    -- get released in case of an exception.
+    expectCompletedChildren ::
+         ActionRegistry m
+      -> PendingMerge m h
+      -> m (MR.TreeMergeType, Vector (Ref (Run m h)))
+    expectCompletedChildren reg (PendingMerge ty prs mts) = do
+        rs1 <- V.forM prs $ \case
+          PreExistingRun r -> do
+            delayedCommit reg (releaseRef r)  -- only released at the end
+            withRollback reg (dupRef r) releaseRef
+          PreExistingMergingRun mr -> do
+            delayedCommit reg (releaseRef mr)  -- only released at the end
+            withRollback reg (MR.expectCompleted mr) releaseRef
```

Comment:
Whereas here it makes sense because we're allocating a run from an old merging run, and then releasing the old merging run.